rustls-symcrypt/src/verify.rs

```diff
@@ -0,0 +1,322 @@
+/// Creates either a RSA or ECC key depending on the SymCryptAlgorithm then does a verify based on the
+/// provided `public_key` bytes, `message` bytes and the `signature` bytes.
+impl SignatureVerificationAlgorithm for SymCryptAlgorithm {
+    fn verify_signature(
+        &self,
+        public_key: &[u8],
+        message: &[u8],
+        signature: &[u8],
+    ) -> Result<(), InvalidSignature> {
+        match &self.key_type {
+            KeyType::Ecc(ecc) => {
+                // the pub_key passed will have a 0x04 legacy byte prepended to the key for the case of
+                // NistP256, NistP384, NistP521.
+                let key = extract_ecc_public_key(public_key, ecc.curve);
+
+                // the signature will be in ASM.1 DER format, with seperated `r` and `s` components, need to remove padding
+                // and concatenate the two components.
+                let sig = extract_ecc_signature(&signature)?;
+
+                let ec_key = EcKey::set_public_key(ecc.curve, &key, EcKeyUsage::EcDsa).unwrap();
+                let hashed_message = (self.hasher)(message);
+                ec_key
+                    .ecdsa_verify(&sig, &hashed_message)
+                    .map_err(|_| InvalidSignature)
+            }
+            KeyType::RsaPkcs1(rsa_pkcs1) => {
+                // extract the modulus and exponent from the public key
+                let (modulus, exponent) = extract_rsa_public_key(public_key)?;
+                let rsa_key =
+                    RsaKey::set_public_key(&modulus, &exponent, RsaKeyUsage::SignAndEncrypt)
+                        .unwrap();
+                let hashed_message = (self.hasher)(message);
+                rsa_key
+                    .pkcs1_verify(&hashed_message, signature, rsa_pkcs1.hash_algorithm)
+                    .map_err(|_| InvalidSignature)
+            }
+            KeyType::RsaPss(rsa_pss) => {
+                // extract the modulus and exponent from the public key
+                let (modulus, exponent) = extract_rsa_public_key(public_key)?;
+                let rsa_key =
+                    RsaKey::set_public_key(&modulus, &exponent, RsaKeyUsage::SignAndEncrypt)
+                        .unwrap();
+                let hashed_message = (self.hasher)(message);
+                rsa_key
+                    .pss_verify(
+                        &hashed_message,
+                        signature,
+                        rsa_pss.hash_algorithm,
+                        rsa_pss.salt_length as usize,
+                    )
+                    .map_err(|_| InvalidSignature)
+            }
+        }
+    }
+
+    fn public_key_alg_id(&self) -> AlgorithmIdentifier {
+        self.public_key_alg_id
+    }
+
+    fn signature_alg_id(&self) -> AlgorithmIdentifier {
+        self.signature_alg_id
+    }
+
+    fn fips(&self) -> bool {
+        // For now, leave fips return as always false.
+        // TODO: investigate fips flag in symcrypt.
```

Comment:
Calling this out, we should turn this on once we remove ring dep. for signer

Reply:
Doesn't this only apply to the above verify functions which now call symcrypt?